Add Sample Documentation (#55)
LM-Development/aks-sample: Samples/PublicSamples/RecordingBot/docs/tutorials/deploy/aks.md

# Deploy an AKS cluster

> We'll start with deploying an AKS cluster, on the AKS cluster we later deploy the containers with the sample recording bot. Before we can start to run commands in the Azure command line tool, we have to login, to do so, we run:

Reviewer: ```suggestion
Start by deploying an AKS cluster. Before you can start to run commands in the Azure CLI tool, you have to login. To do so, run:

```

Author: Diátaxis suggests to use the "we" form, the learner does not go alone threw the tutorial we the teacher and the learner together do a tutorial

```powershell
az login --tenant 99999999-9999-9999-9999-999999999999
```

After running the command, it should show the following message:

```text
The default web browser has been opened at https://login.microsoftonline.com/99999999-9999-9999-9999-999999999999/oauth2/v2.0/authorize. Please continue the login in the web browser. If no web browser is available or if the web browser fails to open, use device code flow with `az login --use-device-code`.
```

And a Browser window with the Microsoft Login Page should open.

![Login Page](../../images/loginscreenshot.png)

> There we login with our Microsoft Entra Id administrator account and accept the scopes requested.

Reviewer: ```suggestion
Log in with your Microsoft Entra Id administrator account and accept the requested scopes.

```

> After successful log in, we should see a output similar to:

Reviewer: ```suggestion
After successful log in, the output should look as follows:

```

```json
[
  {
    "cloudName": "AzureCloud",
    "homeTenantId": "99999999-9999-9999-9999-999999999999",
    "id": "yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyyy",
    "isDefault": true,
    "managedByTenants": [],
    "name": "recordingbotsubscription",
    "state": "Enabled",
    "tenantId": "99999999-9999-9999-9999-999999999999",
    "user": {
      "name": "user@xyz.com",
      "type": "user"
    }
  }
]
```

## Create Azure Resource Group

> Now we can start to create resources in our azure subscription. We start with creating a resource group in our Azure Subscription.

Reviewer: ```suggestion
Create a resource group in your Azure subscription:
```

```powershell
az group create 
    --location westeurope 
    --name recordingbottutorial 
    --subscription "recordingbotsubscription"
```

The result in the command line should look something like:

```json
{
  "id": "/subscriptions/yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyyy/resourceGroups/recordingbottutorial",
  "location": "westeurope",
  "managedBy": null,
  "name": "recordingbottutorial",
  "properties": {
    "provisioningState": "Succeeded"
  },
  "tags": null,
  "type": "Microsoft.Resources/resourceGroups"
}
```

## Create Azure Kubernetes Cluster

> Now we can create an AKS cluster in the resource group, we will create a free tier cluster, this doesn't mean the nodes of the cluster are for free but the managment plane is. A free tier cluster is for testing and development and should not be used for production, see [pricing tiers](https://learn.microsoft.com/en-us/azure/aks/free-standard-pricing-tiers) for reference. And set the node count of the system nodepool, that is automatically created, to 1, because we want to avoid unnecessary cost from this tutorial. For the size of the system nodes in the system nodepool we choose the `standard_d2s_v3`-series, as this series is available at the most regions and some nodes in this series are available without requesting more quotas. Still you might need to check you're quotas, see [per vm quotas](https://learn.microsoft.com/en-us/azure/quotas/per-vm-quota-requests) for reference.

Reviewer: ```suggestion
Let's get to creating the AKS cluster in the resource group! Create a free tier cluster (This doesn't mean the nodes of the cluster are for free but the management plane is. A free tier cluster is for testing and development and should not be used for production, see [pricing tiers](https://learn.microsoft.com/en-us/azure/aks/free-standard-pricing-tiers) for reference). Set the node count of the automatically created system nodepool to 1, because we want to avoid unnecessary cost from this tutorial. Choose the `standard_d2s_v3`-series, as this series is available at most regions and some nodes in this series are available without requesting additional quotas. Still you might need to check your quotas, see [per vm quotas](https://learn.microsoft.com/en-us/azure/quotas/per-vm-quota-requests) for reference.

```

Reviewer 2: I still believe there should be a warning block for possibly incurring costs
(its easy to overlook this detail in a block of text, especially if you're just trying to get something done quickly)

> So the command we exectute to deploy our AKS cluster we run the command:

Reviewer: ```suggestion
Run the following command with your settings:

```

  "powerState": {
    "code": "Running"
  },
  "privateFqdn": null,
  "privateLinkResources": null,
  "provisioningState": "Succeeded",
  "publicNetworkAccess": null,
  "resourceGroup": "recordingbottutorial",
  "resourceUid": "0123456789abcdef12345678",
  "securityProfile": {
    "azureKeyVaultKms": null,
    "defender": null,
    "imageCleaner": null,
    "workloadIdentity": null
  },
  "serviceMeshProfile": null,
  "servicePrincipalProfile": {
    "clientId": "msi",
    "secret": null
  },
  "sku": {
    "name": "Base",
    "tier": "Free"
  },
  "storageProfile": {
    "blobCsiDriver": null,
    "diskCsiDriver": {
      "enabled": true
    },
    "fileCsiDriver": {
      "enabled": true
    },
    "snapshotController": {
      "enabled": true
    }
  },
  "supportPlan": "KubernetesOfficial",
  "systemData": null,
  "tags": null,
  "type": "Microsoft.ContainerService/ManagedClusters",
  "upgradeSettings": null,
  "windowsProfile": {
    "adminPassword": null,
    "adminUsername": "azureuser",
    "enableCsiProxy": true,
    "gmsaProfile": null,
    "licenseType": null
  },
  "workloadAutoScalerProfile": {
    "keda": null,
    "verticalPodAutoscaler": null
  }
}
```

> [!NOTE]  

> > As we can see in the result json we created with the creation of the AKS cluster some more resources, e.g. a virtual machine scale set, a public IP and more. These resources are in a newly created resource group, in our case this resource group is called `MC_recordingbottutorial_recordingbotcluster_westeurope`. Search in the json how this resource group is called in your case, it should be in a similar pattern and write down this name as we need it later, to write it down now will save you some time later.

Reviewer: ```suggestion
> As you can see in the result json, the creation of the AKS cluster also created some additional resources, e.g. a virtual machine scale set, a public IP and more. These auxiliary resources are in a newly created resource group, in our case this resource group is called `MC_recordingbottutorial_recordingbotcluster_westeurope`. Search in the json how this resource group is called in your case, it should be in a similar pattern and write down this name as we need it later, to write it down now will save you some time later.

```

  "kubeletConfig": null,
  "kubeletDiskType": "OS",
  "linuxOsConfig": null,
  "maxCount": null,
  "maxPods": 30,
  "minCount": null,
  "mode": "System",
  "name": "nodepool1",
  "networkProfile": null,
  "nodeImageVersion": "AKSUbuntu-2204gen2containerd-202403.25.0",
  "nodeLabels": null,
  "nodePublicIpPrefixId": null,
  "nodeTaints": null,
  "orchestratorVersion": "1.28.5",
  "osDiskSizeGb": 128,
  "osDiskType": "Managed",
  "osSku": "Ubuntu",
  "osType": "Linux",
  "podSubnetId": null,
  "powerState": {
    "code": "Running"
  },
  "provisioningState": "Succeeded",
  "proximityPlacementGroupId": null,
  "resourceGroup": "recordingbottutorial",
  "scaleDownMode": null,
  "scaleSetEvictionPolicy": null,
  "scaleSetPriority": null,
  "spotMaxPrice": null,
  "tags": null,
  "type": "Microsoft.ContainerService/managedClusters/agentPools",
  "typePropertiesType": "VirtualMachineScaleSets",
  "upgradeSettings": {
    "drainTimeoutInMinutes": null,
    "maxSurge": "10%",
    "nodeSoakDurationInMinutes": null
  },
  "vmSize": "standard_d2s_v3",
  "vnetSubnetId": null,
  "workloadRuntime": null
}
```

As you can see the _nodeTaints_ field now updated to `null`.

## Set DNS name

> When we created our AKS cluster, we automatically created a public IP too. Now we need to create a DNS name for the public IP. The resouce of the IP is not in our default resource group `recordingbottutorial`, because the creation of our AKS cluster created a Resource group with the managed resources. If you already know the resource group name of the managed resources, you can coninue with [getting the public IP resource name](#get-the-public-ip-resource-name).

Reviewer 2: ```suggestion
When we created our AKS cluster, a public IP was automatically created for us. To use this IP Address, we want to create a DNS entry so we can reference it by name. Since the IP Address was automatically created for us, it resides in the automatically created resource group which is an Azure Managed Resource. If you have not noted down the name of this resource group in [Create Azure Kubernetes Cluster](#create-azure-kubernetes-Cluster), follow the next step to find out the name. Otherwise you can skip ahead to [getting the public IP resource name](#get-the-public-ip-resource-name).

```

### Get the managed resources resource group name

We can find the managed resource group in the description of the AKS cluster:

```powershell
az aks show
    --resource-group recordingbottutorial
    --name recordingbotcluster
    --subscription "recordingbotsubscription" | Select-String 
            -Pattern 'nodeResourceGroup' 
            -SimpleMatch
            -NoEmphasis
```

This will give us the response

```json
  "nodeResourceGroup": "MC_recordingbottutorial_recordingbotcluster_westeurope",
```

> as the node resource group is the resource group where our cluster manges the resources, in our case the resource group is called `MC_recordingbottutorial_recordingbotcluster_westeurope`.

Reviewer 2: ```suggestion
The IP Address Resource was created in the Azure Managed Resource Group which is used by Azure to manage the AKS Cluster. You can find the name in the _nodeResourceGroup_ field of the output. In the example output the resource group is called `MC_recordingbottutorial_recordingbotcluster_westeurope`.

```

### Get the public IP resource name

> Next we need the resource name of the public IP. To get the name we list the public IP resources in the managed resource group.

Reviewer 2: ```suggestion
Since the Public IP Address is an Azure Resource, it can be referenced by name. To find out, what name was assigned to it when it, execute the following command:

```

```powershell
az network public-ip list 
    --resource-group MC_recordingbottutorial_recordingbotcluster_westeurope
    --subscription "recordingbotsubscription"
```

> The resulting list

Reviewer 2: Consider the following example output:

    "idleTimeoutInMinutes": 4,
    "ipAddress": "108.141.184.42",
    "ipConfiguration": {
      "id": "/subscriptions/yyyyyyyy-yyyy-yyyy-yyyy-yyyyyyyyyyyyy/resourceGroups/MC_recordingbottutorial_recordingbotcluster_westeurope/providers/Microsoft.Network/loadBalancers/kubernetes/frontendIPConfigurations/cab190bb-ec74-478e-b7f1-b36c83bfa94e",
      "resourceGroup": "MC_recordingbottutorial_recordingbotcluster_westeurope"
    },
    "ipTags": [],
    "location": "westeurope",
    "name": "cab190bb-ec74-478e-b7f1-b36c83bfa94e",
    "provisioningState": "Succeeded",
    "publicIPAddressVersion": "IPv4",
    "publicIPAllocationMethod": "Static",
    "resourceGroup": "MC_recordingbottutorial_recordingbotcluster_westeurope",
    "resourceGuid": "539f439e-5c50-4bc6-a4a9-fb2d102e88f3",
    "sku": {
      "name": "Standard",
      "tier": "Regional"
    },
    "tags": {
      "aks-managed-cluster-name": "recordingbotcluster",
      "aks-managed-cluster-rg": "recordingbottutorial",
      "aks-managed-type": "aks-slb-managed-outbound-ip"
    },
    "type": "Microsoft.Network/publicIPAddresses",
    "zones": [
      "3",
      "1",
      "2"
    ]
  }
]
```

> The list only has one element as our AKS cluster only created one public IP, in our case the name of the public IP is `cab190bb-ec74-478e-b7f1-b36c83bfa94e`.

Reviewer 2: ```suggestion
The name of the IP Address Resource can be found in the _name_ field of the output. In the example output the name of the Public IP Address Resource is `cab190bb-ec74-478e-b7f1-b36c83bfa94e`.

```

### Set DNS name for public IP resource

> With the managed resource group name and the public IP resource name, we can now set a DNS name for the public IP resource with the command:

Reviewer 2: ```suggestion
With the managed resource group name and the Public IP Address resource name, we can now set a DNS name for the Public IP Address resource with the command:

```

Author: Why do you want to write `public IP address` in upper case?

Reviewer 2: cause i think it was done somewhere else (in this document) and we should stay consistent
it does make sense to me, since its the "name" of a resource
but its fine in eng to use lower case

It can also be used as a means to highlight certain terms though

Author: Throughout this tutorial it is written lowercase everywhere now, i would choose consitency now

```powershell
az network public-ip update 
    --resource-group MC_recordingbottutorial_recordingbotcluster_westeurope 
    --name cab190bb-ec74-478e-b7f1-b36c83bfa94e
    --dns-name recordingbottutorial
    --subscription "recordingbotsubscription"
```

Don't forget to replace the DNS name with your own AKS DNS record, only the variable part, for example we chose `recordingbottutorial`_.westeurope.cloudapp.azure.com_ so for the `--dns-name` parameter we enter `recordingbottutorial`.

> [!WARNING]  
> The DNS record that is created from our custom part and the postfix must be globally unique.

> If the command ran successful the result will look like:

Reviewer 2: ```suggestion
If the command completed successfully, the result will look similar to this:

```

  "provisioningState": "Succeeded",
  "publicIPAddressVersion": "IPv4",
  "publicIPAllocationMethod": "Static",
  "resourceGroup": "MC_recordingbottutorial_recordingbotcluster_westeurope",
  "resourceGuid": "539f439e-5c50-4bc6-a4a9-fb2d102e88f3",
  "sku": {
    "name": "Standard",
    "tier": "Regional"
  },
  "tags": {
    "aks-managed-cluster-name": "recordingbotcluster",
    "aks-managed-cluster-rg": "recordingbottutorial",
    "aks-managed-type": "aks-slb-managed-outbound-ip"
  },
  "type": "Microsoft.Network/publicIPAddresses",
  "zones": [
    "3",
    "1",
    "2"
  ]
}
```

As we can see a field was added to the resource called `dnsSettings`. Within the DNS settings we can see the custom part of our DNS record, that we came up with, and the fully qualified domain name (fqdn) with the postfix provided by Azure.

If the fqdn already exists we would get the follwing error message.

```text
(DnsRecordInUse) DNS record recordingbottutorial.westeurope.cloudapp.azure.com is already used by another public IP.
Code: DnsRecordInUse
Message: DNS record recordingbottutorial.westeurope.cloudapp.azure.com is already used by another public IP.
```

> then we have to come up with a new prefix for the DNS record.

Reviewer 2: ```suggestion
then we would have to come up with a new prefix for the DNS record.

```

We can install kubectl with the Azure command line tool:

```powershell
az aks install-cli
```

The resulting output should look similar to

```text
The detected architecture of current device is "amd64", and the binary for "amd64" will be downloaded. If the detection is wrong, please download and install the binary corresponding to the appropriate architecture.
No version specified, will get the latest version of kubectl from "https://storage.googleapis.com/kubernetes-release/release/stable.txt"
Downloading client to "C:\Users\User\.azure-kubectl\kubectl.exe" from "https://storage.googleapis.com/kubernetes-release/release/v1.29.4/bin/windows/amd64/kubectl.exe"
No version specified, will get the latest version of kubelogin from "https://api.github.com/repos/Azure/kubelogin/releases/latest"
Downloading client to "C:\Users\User\AppData\Local\Temp\tmp56tfm9jk\kubelogin.zip" from "https://github.com/Azure/kubelogin/releases/download/v0.1.1/kubelogin.zip"
Moving binary to "C:\Users\User\.azure-kubelogin\kubelogin.exe" from "C:\Users\User\AppData\Local\Temp\tmp56tfm9jk\bin\windows_amd64\kubelogin.exe"
```

> Now we installed the command line tools for our kubernetes cluster, we might need to restart our powershell.

Reviewer 2: ```suggestion
> [!NOTE]
> After installing kubectl on our kubernetes cluster, we might need to open a new powershell session. We probably need to log in again as described in [Deploy an AKS cluster](#deploy-an-aks-cluster).

```
